### ISSUE TITLE
Bug in handling partial data

### ISSUE THREAD

**Issue by Reiner030** [ggrandes/apache22-modules#1]
Hello @ggrandes,

I tried out your nice module apache22-modules/mod_myfixip.c but it's not working as expected in my environment. Test against Nginx works fine with my AWS ELB ProxyProtocol setup.

1) After longer digging I found out that the `apr_bucket_read()` gives back a wrong length.
    Every time only 11 chars instead of 46/47.
    Thats only the part "PROXY TCPv4 SRCIP DSTIP SRCPORT DSTPORT ":

```
414     // Read first bucket
415     apr_bucket_read(e, &str, &length, APR_BLOCK_READ);
416 #ifdef DEBUG
417         ap_log_error(APLOG_MARK, APLOG_WARNING, 0, NULL, MODULE_NAME "::helocon_filter_in DEBUG: header='%s', read length=%d, real length=%d", str, length, strlen(str));
418 #endif
```

2) there is still no successfull connection possible - I guess that there could be still a buffer problem open which breaks the SSL handshake. Do you have perhaps an idea what could happened?
I am using Debian Wheezy with apache2.2 `2.2.22-13+deb7u3` / now `2.2.22-13+deb7u4.`.

Bests

**Comment by ggrandes** [comment]
Currently I have no idea, but... I will investigate this.
Thanks.

**Comment by ggrandes** [comment]
I have reproduced the problem, I hope find a solution shortly.

**Comment by ggrandes** [comment]
Fixed [#1] [commit be8f0f00ac27c47ef6f3dc92e02809a609cd36d0]

**Comment by ggrandes** [comment]
The problem was to assume that the data would arrive together in only one bucket.
From [PROXY protocol](http://www.haproxy.org/download/1.5/doc/proxy-protocol.txt):

> Both formats are designed to fit in the smallest TCP segment that any TCP/IP
> host is required to support (576 - 40 = 536 bytes). This ensures that the whole
> header will always be delivered at once when the socket buffers are still empty
> at the beginning of a connection. The sender must always ensure that the header
> is sent at once, so that the transport layer maintains atomicity along the path
> to the receiver.

**Comment by Reiner030** [comment]
Big THX.., now PROXY function works with version 0.6 nice:

```
192.168.45.16 - - [17/Jan/2015:16:41:00 +0100] "GET / HTTP/1.1" 200 2004 "-" "lwp-request/6.03 libwww-perl/6.04"
192.168.45.2 - - [17/Jan/2015:16:41:41 +0100] "GET / HTTP/1.1" 200 2017 "-" "fetch libfetch/2.0"
```

I would prepare then the connections to switch to the nicer PROXY protocol because it supports SSL SNI and SPDY usage ;)

(btw: an improved old/debian stable available mod_rpaf to use also Proxy IP areas, IPv6  by backported partly from mod-remote-ip functions of 2.4 is here: https://github.com/gnif/mod_rpaf )

**Comment by ggrandes** [comment]
Fixed really in [c35ca38466f867eca35d1b9c3ecc718e5f0f3142]

**Comment by ggrandes** [comment]
I think that I don't fully understand... What do you want to do?

**Comment by Reiner030** [comment]
a small "bug" for Debugging: readbytes is formatted as %lld not %11d:

```
/files/mod_myfixip.c.orig  2015-01-17 16:36:23.216287915 +0100
+++ salt/base/apache2/files/mod_myfixip.c       2015-01-23 12:01:36.045701072 +0100
@@ -381,7 +381,7 @@
     apr_bucket *e = NULL;

 #ifdef DEBUG
-    ap_log_error(APLOG_MARK, APLOG_WARNING, 0, NULL, MODULE_NAME "::helocon_filter_in IP Connection from: %s to port=%d readbytes=%lld (1)", _CLIENT_IP, c->local_addr->port, readbytes);
+    ap_log_error(APLOG_MARK, APLOG_WARNING, 0, NULL, MODULE_NAME "::helocon_filter_in IP Connection from: %s to port=%d readbytes=%11d (1)", _CLIENT_IP, c->local_addr->port, readbytes);
 #endif

     // Fail quickly if the connection has already been aborted.
@@ -498,7 +498,7 @@
                 // Refill buckets
                 if ((e == APR_BRIGADE_SENTINEL(b)) && APR_BRIGADE_EMPTY(b)) {
 #ifdef DEBUG
-                    ap_log_error(APLOG_MARK, APLOG_WARNING, 0, NULL, MODULE_NAME "::helocon_filter_in DEBUG: refill buffer size=%lld (1)", readbytes);
+                    ap_log_error(APLOG_MARK, APLOG_WARNING, 0, NULL, MODULE_NAME "::helocon_filter_in DEBUG: refill buffer size=%11d (1)", readbytes);
 #endif
                     ap_get_brigade(f->next, b, mode, block, readbytes);
                     e = APR_BRIGADE_FIRST(b);
@@ -570,7 +570,7 @@
         // Refill buckets
         if (APR_BRIGADE_EMPTY(b)) {
 #ifdef DEBUG
-            ap_log_error(APLOG_MARK, APLOG_WARNING, 0, NULL, MODULE_NAME "::helocon_filter_in DEBUG: refill buffer size=%lld (2)", readbytes);
+            ap_log_error(APLOG_MARK, APLOG_WARNING, 0, NULL, MODULE_NAME "::helocon_filter_in DEBUG: refill buffer size=%11d (2)", readbytes);
 #endif
             ap_get_brigade(f->next, b, mode, block, readbytes);
             e = APR_BRIGADE_FIRST(b);
```

(I have today again AWS buffer overflow problems :( )

**Comment by ggrandes** [comment]
readbytes is apr_off_t

from /usr/include/apr-1.0/apr.h:

``` c
    typedef  off64_t           apr_off_t;
    #define APR_OFF_T_FMT APR_INT64_T_FMT
    #define APR_INT64_T_FMT "lld"
    #define APR_OFF_T_FMT "lld"
```

So, I used "lld", but yes, this is not 100% correct. 
Better I use defined format string. Fixed [99674b46dfde53bb05d78d5cddf051946bb7f8fa]

**Comment by ggrandes** [comment]
you can tell me about these overflows?

**Comment by ggrandes** [comment]
@Reiner030 the latest version (1.2) fix all problems :-)

**Comment by Reiner030** [comment]
Hi @ggrandes , patch looks nice/that there are the right function calls found...
Sadly I have no access to the setup anymore but I would inform the new devs that they can now update/check the AWS ELB in proxy mode. 
But I can try/test it on a new setup to verify it - just checked that same fix is also made in 2.4 module ..  Thanx ;)

**Comment by ggrandes** [comment]
:+1:
